### PR TITLE
BBCode

### PR DESCRIPTION
**Pull request by cod1r**

### CODE REVIEW

File: lib/xdparser2.ts [hunk 1]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
-      if (letter === "~") {
-        mode = "text"
-        components.push(["strike", token])
-        token = ""
-        continue
+  if (textSlice.length > 0) {
+    components.push(["text", textSlice])
+  }
+
+  const urlRegex = new RegExp(/\[([a-zA-Z ]*)\]\(([a-zA-Z0-9\/:\.\?\&\%\$\!\@\#\-\+\~\,\= ]*)\)/g)
```
Comment (samanpwbb) [inline lib/xdparser2.ts:748]: Can this regex could be simplified to allow any character inside ()?

File: lib/xdparser2.ts [hunk 2]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
+  let i = 0
+  while (i < components.length) {
+    const component = components[i]
+    if (component[0] === "text") {
+      const matches = component[1].matchAll(urlRegex)
+      const length = component[1].length
+      component[1].replace(urlRegex, '')
```
Comment (samanpwbb) [inline lib/xdparser2.ts:755]: Is this a replace on a string? If so, this line isn't doing anything? Shouldn't it be:

component[1] = component[1].replace(urlRegex,'')

File: lib/xdparser2.ts [hunk 3]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
 
-      // Bold is started by two asterisks
-      if (letter === "*" && nextLetter === "*") {
-        mode = "bold"
-        // Jump a letter
-        index++
-        pushText()
-        continue
+    if (mdTokens.includes(slice)) {
+      // tries the find the most recent operator that matches the current one
```
Comment (samanpwbb) [inline lib/xdparser2.ts:709]: Does it make sense to apply this approach to the link keys ("[", "]", "(", ")")? Aren't you already doing link parsing with a regex further below?

File: lib/xdparser2.ts [hunk 4]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
-    } else if (mode === "text" && prevLetter !== "\\") {
-      if (letter === "[") {
-        mode = "link"
-        pushText()
-        continue
+  const escapeCharacter = "^"
+  let textSlice = ''
+  // stack of [index, md operators like "**" etc]
+  const stack: Array<[number, string]> = []
+  for (let index = 1; index < str.length; index++) {
```
Comment (samanpwbb) [inline lib/xdparser2.ts:693]: If you are going through the effort or rewriting this parser, go one more step further and use some functions. It will help the next person understand the purpose of different parts of this code.

File: lib/xdparser2.ts [hunk 5]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
-    if (letter === "\\" && otherMDKeys.includes(nextLetter)) {
-      index++
-      token += nextLetter
-      continue
-    } else if (mode === "text" && prevLetter !== "\\") {
-      if (letter === "[") {
-        mode = "link"
-        pushText()
-        continue
+  const escapeCharacter = "^"
```
[the remark this answers is not in the document]
Comment (samanpwbb) [reply lib/xdparser2.ts:689]: this ^ character is already used for hint metadata (https://github.com/puzzmo-com/xd-crossword-tools?tab=readme-ov-file#clue-metadata)

File: lib/xdparser2.ts [hunk 6]
```diff
@@ -676,92 +676,104 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
-
-  const pushText = () => {
-    if (token.length > 0) {
-      components.push(["text", token])
-      token = ""
+  const mdTokens = ["**", "\\\\", "~~", "[", "]", "(", ")"]
+  const getMDType = (operator: string) => {
+    switch (operator) {
+      case "**": return "bold"
+      case "\\\\": return "italics"
```
[the remark this answers is not in the document]
Comment (orta) [reply lib/xdparser2.ts:683]: Long story, but it should be `//` not `\\` kinda interested to know where that came from too - looks like a discussion in here which I'm also gonna be blocking these changes on: https://github.com/puzzmo-com/games/issues/1709

it should be markdown-ish but the doubles trigger formatting for backwards compatibility

[the remark this answers is not in the document]
Comment (orta) [reply lib/xdparser2.ts:683]: Seems to be that `//` appears naturally in `https://www.` - so I think we make the parser always require the space before to start for all tokens

[the remark this answers is not in the document]
Comment (cod1r) [reply lib/xdparser2.ts:683]: yea i'll mention this in the issue. i'll change things to do `//`

[the remark this answers is not in the document]
Comment (samanpwbb) [reply lib/xdparser2.ts:683]: BBCode would be better than HTML, either would be better than what we have now. BBCode is superior because we don't have to worry about escaping it crossword editor where we may want to display the contents of an xd file. Another benefit of BBCode is that it is the least likely to collide with a crossword constructor using the the syntax unintentionally in a clue somewhere (HTML probably also safe in this case though). So, in BBCode style, we'd have:

- [b]{text}[/b]
- [i]{text}[/i]
- [s]{text}[/s]
- [url={url}]{text}[/url]

From https://www.bbcode.org/reference.php

[the remark this answers is not in the document]
Comment (cod1r) [reply lib/xdparser2.ts:683]: ok this is cool. first time i've seen bbcode 👍

**Comment by cod1r** [comment]
I really like the BBCode syntax.

**Comment by cod1r** [comment]
this is ready. i didn't get the version right and orta wanted the do something with the docs

File: lib/xdparser2.ts [hunk 7]
```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
   return splits
 }
 
-export function inlineMarkdownParser(str: string): MDClueComponent[] {
-  const components: MDClueComponent[] = []
-  let token = ""
-  let mode: MDClueComponent[0] = "text"
-  let linkText = ""
+function isAlphabetical(str: string): boolean {
```
Comment (samanpwbb) [inline lib/xdparser2.ts:677]: This name is confusing, maybe, isFirstCharAlphabetic?

File: lib/xdparser2.ts [hunk 8]
```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
+  return str.charCodeAt(0) >= 97 && str.charCodeAt(0) <= 122
```
Comment (samanpwbb) [inline lib/xdparser2.ts:678]: This would be clearer if you just used a regex (/^[a-z]/.test(str)). But should this actually be more permissive than just checking a-z? What about diacritics, etc.

File: lib/xdparser2.ts [hunk 9]
```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
+}
 
-      if (letter === "/" && prevLetter === " ") {
-        mode = "italics"
-        pushText()
-        continue
-      }
+function parseBBInnerContents(str: string) {
+  let index = 0;
+  let start = index;
```
Comment (samanpwbb) [inline lib/xdparser2.ts:719]: what does this start variable do? Seems like the value is just zero.

File: lib/xdparser2.ts [hunk 10]
```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
-    } else if (mode === "strike") {
-      if (letter === "~") {
-        mode = "text"
-        components.push(["strike", token])
-        token = ""
+    }
+  }
+}
+
+export function inlineMarkdownParser(str: string): MDClueComponent[] {
```
Comment (samanpwbb) [inline lib/xdparser2.ts:778]: rename this function

File: lib/xdparser2.ts [hunk 11]
```diff
@@ -674,94 +674,130 @@ function parseSplitsFromAnswer(answerWithSplits: string, splitCharacter?: string
...
+  let starting = 0 // used for keeping the start of the entire bb tag/group
+  let tagType = parseBBTagType(str.slice(index))
+  if (tagType === "b" || tagType === "url" || tagType === "i" || tagType === "s") {
+    index += tagType.length
+    let url;
+    if (tagType === "url") {
+      url = parseURL(str.slice(index))
+      if (url.length === 0) {
+        // no url so invalid tag
+        return
```
Comment (samanpwbb) [inline lib/xdparser2.ts:737]: What happens if parseBB returns undefined exactly? Could be more explicit to throw an error and handle the error in the caller.

File: tests/xdparser2/xdparser.mdclue.test.ts [hunk 12]
```diff
@@ -40,139 +50,92 @@ it("handles bolding", () => {
...
-})
 
-it("handles URLs", () => {
-  const xd = readFileSync("tests/xdparser2/inputs/alpha-bits.xd", "utf8")
-  const originalClue = "A1. Captain of the Pequod ~ AHAB"
-  const newMDClue = "A1. [Captain](https://github.com/orta) **of** the ship /Pequod/ ~ AHAB"
+})
 
-  expect(xdParser(xd.replace(originalClue, newMDClue)).clues.across[0].bodyMD).toMatchInlineSnapshot(`
+it("handles links, bolds, italics, strikes, and dates", () => {
```
Comment (samanpwbb) [inline tests/xdparser2/xdparser.mdclue.test.ts:67]: this test is fine, but consider also adding individual tests for each tag types, and try to cover more edge cases

**Review by samanpwbb — APPROVED** [review]
This looks good, left a few notes but mergeable in it's current state.

File: lib/xdparser2.ts (hunk 8 above)
Comment (cod1r) [reply lib/xdparser2.ts:678]: eh i think this is fine. i can always change if we need to be more permissive

File: lib/xdparser2.ts (hunk 11 above)
Comment (cod1r) [reply lib/xdparser2.ts:737]: we could throw but something failing to parse as BBCode would be just added as ur typical text so 🤷

File: tests/xdparser2/xdparser.mdclue.test.ts (hunk 12 above)
Comment (cod1r) [reply tests/xdparser2/xdparser.mdclue.test.ts:136]: ok

**Comment by cod1r** [comment]
alright, im merging
